main.go

```diff
@@ -0,0 +1,252 @@
+
+infosquito:
+  maximum_in_prefix: 10000
+  base_prefix_length: 3
+
+elasticsearch:
+  base: http://elasticsearch:9200
+  index: data
+
+db:
+  uri: postgres://ICAT:fakepassword@icat-db:5432/ICAT?sslmode=disable
+`
+
+var log = logrus.WithFields(logrus.Fields{
+	"service": "infosquito2",
+	"art-id":  "infosquito2",
+	"group":   "org.cyverse",
+})
+
+var (
+	cfgPath = flag.String("config", "", "Path to the configuration file.")
+	mode    = flag.String("mode", "", "One of 'periodic' or 'full'.")
+	debug   = flag.Bool("debug", false, "Set to true to enable debug logging")
+	cfg     *viper.Viper
+
+	listenClient  *messaging.Client
+	publishClient *messaging.Client
+
+	amqpURI               string
+	amqpExchangeName      string
+	amqpExchangeType      string
+	amqpQueuePrefix       string
+	elasticsearchBase     string
+	elasticsearchUser     string
+	elasticsearchPassword string
+	elasticsearchIndex    string
+	dbURI                 string
+	maxInPrefix           int
+	basePrefixLength      int
+)
+
+func init() {
+	flag.Parse()
+	logrus.SetFormatter(&logrus.JSONFormatter{})
+	if !(*debug) {
+		logrus.SetLevel(logrus.InfoLevel)
+	} else {
+		logrus.SetLevel(logrus.DebugLevel)
+	}
+}
+
+func spin() {
+	spinner := make(chan int)
+	for {
+		select {
+		case <-spinner:
+			fmt.Println("Exiting")
+			break
+		}
+	}
+}
+
+func checkMode() {
+	if *mode != "periodic" && *mode != "full" {
+		fmt.Printf("Invalid mode: %s\n", *mode)
+		flag.PrintDefaults()
+		os.Exit(-1)
+	}
+}
+
+func initConfig(cfgPath string) {
+	var err error
+	cfg, err = configurate.InitDefaults(cfgPath, defaultConfig)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	dbURI = cfg.GetString("db.uri")
+	elasticsearchBase = cfg.GetString("elasticsearch.base")
+	elasticsearchUser = cfg.GetString("elasticsearch.user")
+	elasticsearchPassword = cfg.GetString("elasticsearch.password")
+	elasticsearchIndex = cfg.GetString("elasticsearch.index")
+	max, err := strconv.Atoi(cfg.GetString("infosquito.maximum_in_prefix"))
+	if err != nil {
+		log.Fatal("Couldn't parse integer out of infosquito.maximum_in_prefix")
+	}
+	maxInPrefix = max
+
+	base, err := strconv.Atoi(cfg.GetString("infosquito.base_prefix_length"))
+	if err != nil {
+		log.Fatal("Couldn't parse integer out of infosquito.base_prefix_length")
+	}
+	basePrefixLength = base
+}
+
+func loadAMQPConfig() {
+	amqpURI = cfg.GetString("amqp.uri")
+	amqpExchangeName = cfg.GetString("amqp.exchange.name")
+	amqpExchangeType = cfg.GetString("amqp.exchange.type")
+	amqpQueuePrefix = cfg.GetString("amqp.queue_prefix")
+}
+
+func getQueueName(prefix string) string {
+	if len(prefix) > 0 {
+		return fmt.Sprintf("%s.infosquito2", prefix)
+	}
+	return "infosquito2"
+}
+
+func generatePrefixes(length int) []string {
+	prefixes := int(math.Pow(16, float64(length)))
+	res := make([]string, prefixes)
+	for i := 0; i < prefixes; i++ {
+		res[i] = fmt.Sprintf("%0"+strconv.Itoa(length)+"x", i)
+	}
+	return res
+}
+
+func splitPrefix(prefix string) []string {
```

Comment:
This would stop working if we got up to 8, but we can deal with that when it becomes relevant.

Reply:
This function will generate prefixes that match nothing if it's passed an 8-character prefix, because the ninth character of a UUID is `-`, not a hexadecimal digit.

Reply:
The same applies to `generatePrefixes` as well. But it'll be quite a while (possibly effectively forever) before we should actually need prefixes anywhere near that long, so I figured I'd leave it for now.